# Intro to Web Scraping! 

![](https://media.giphy.com/media/1dcWGdOBg0wcU/giphy.gif)

    
### Today's Goals:
1. Retrieve the HTML of a webpage with the `requests` library.
2. Introduction to the `tree` structure of `HTML`.
3. Use the `inspect` tool to sift through the HTML.
4. Parse HTML with the `BeautifulSoup` library.
5. Store data in a `csv` file using the `Pandas` library.

# Let's scrape some data!
The data we are scraping today will be from the [Quotes to Scrape](http://quotes.toscrape.com/) website.


## Step 1:
> **Import the necessary tools for our project**

![](https://media.giphy.com/media/KcE7Dq5f8TTXzZ1LAF/giphy.gif)


In [1]:
# Webscraping
import requests
from bs4 import BeautifulSoup

# Data organization
import pandas as pd

# Visualization
import matplotlib.pyplot as plt
plt.rcParams.update({'font.size': 22})

## Step 2
> **We use the `requests` library to connect to the website we wish to scrape.**

<img src='https://media.giphy.com/media/eCwAEs05phtK/giphy.gif' width='300'></img>

In [2]:
url = 'http://quotes.toscrape.com'
response = requests.get(url)
response

<Response [200]>

✅**A `Response 200` means our request was sucessful!** ✅

❌Let's take a quick look at an *unsuccessful* response. ❌

In [3]:
bad_url = 'http://quotes.toscrape.com/20'
requests.get(bad_url)

<Response [404]>

**A `Response 404` means that the url you are using in your request is not pointing to a valid webpage.**

<img src='https://media.giphy.com/media/VwoJkTfZAUBSU/giphy.gif' width='300'></img>

## Step 3
> **We collect the html from the website by adding `.text` to the end of the response variable.** 

In [5]:
html = response.text
html[:300]

'<!DOCTYPE html>\n<html lang="en">\n<head>\n\t<meta charset="UTF-8">\n\t<title>Quotes to Scrape</title>\n    <link rel="stylesheet" href="/static/bootstrap.min.css">\n    <link rel="stylesheet" href="/static/main.css">\n</head>\n<body>\n    <div class="container">\n        <div class="row header-box">\n          '

## Step 4
> **We use `BeautifulSoup` to turn the html into something we can manipulate.**

In [6]:
soup = BeautifulSoup(html, 'lxml')
soup

<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<title>Quotes to Scrape</title>
<link href="/static/bootstrap.min.css" rel="stylesheet"/>
<link href="/static/main.css" rel="stylesheet"/>
</head>
<body>
<div class="container">
<div class="row header-box">
<div class="col-md-8">
<h1>
<a href="/" style="text-decoration: none">Quotes to Scrape</a>
</h1>
</div>
<div class="col-md-4">
<p>
<a href="/login">Login</a>
</p>
</div>
</div>
<div class="row">
<div class="col-md-8">
<div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
<span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
<span>by <small class="author" itemprop="author">Albert Einstein</small>
<a href="/author/Albert-Einstein">(about)</a>
</span>
<div class="tags">
            Tags:
            <meta class="keywords" content="change,deep-thoughts,thinking,world" itemprop="keywords"/>
<a class="t

<center><h1><b><i><u>Very Soupy</u></i></b></h1></center>

The name ***Beautiful Soup*** is an appropriate description. HTML does not make for a lovely reading experience. If you feel like you're staring at complete gibberish, you're not entirely wrong! HTML is a language designed for computers, not for human eyes.

<img src='https://media.giphy.com/media/5xtDarBbqdSQxfGFdNS/giphy.gif' width="200"></img>

**Fortunately for us,** <u>we do not have to read through every line of the html in order to web scrape.</u> 

Modern day web browsers come equipped with tools that allow us to easily sift through this soupy text.


## Step 4
>**We open up the page we're trying to scrape in a new tab.** <b><a href='http://quotes.toscrape.com/' target='_blank'>Click Here!</a></b> 👀

## Step 5
   > **We create a list of every `div` that has a `class` of "quote"**

**In this instance,** every item we want to collect is divided into identically labeled containers.
- A div with a class of 'quote'.

Not all HTML is as well organized as this page, but HTML is basically just a bunch of different organizational containers that we use to divide up text and other forms of media. Figuring out the organizational structure of a website is the entire process for web scraping. 

In [7]:
quote_divs = soup.find_all('div', {'class': 'quote'})
len(quote_divs)

10

In [8]:
quote_divs[0]

<div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
<span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
<span>by <small class="author" itemprop="author">Albert Einstein</small>
<a href="/author/Albert-Einstein">(about)</a>
</span>
<div class="tags">
            Tags:
            <meta class="keywords" content="change,deep-thoughts,thinking,world" itemprop="keywords"/>
<a class="tag" href="/tag/change/page/1/">change</a>
<a class="tag" href="/tag/deep-thoughts/page/1/">deep-thoughts</a>
<a class="tag" href="/tag/thinking/page/1/">thinking</a>
<a class="tag" href="/tag/world/page/1/">world</a>
</div>
</div>

## Step 6
> **To figure out how to grab all the datapoints from a quote, we isolate a single quote, and work through the code for a single `div`.**

In [9]:
first_quote = quote_divs[0]

### First we grab the text of the quote

In [10]:
text = first_quote.find('span', {'class':'text'})
text = text.text
text

'“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”'

### Next we grab the author

In [11]:
author = first_quote.find('small', {'class': 'author'})
author_name = author.text
author_name

'Albert Einstein'

### Let's also grab the link that points to the author's bio!

In [12]:
author_link = author.findNextSibling().attrs.get('href') #.get means code continues to run even despite errors
author_link = url + author_link
author_link

'http://quotes.toscrape.com/author/Albert-Einstein'

### And finally, let's grab all of the tags for the quote

In [13]:
tag_container = first_quote.find('div', {'class': 'tags'})
tag_links = tag_container.find_all('a')

tags = []
for tag in tag_links:
    tags.append(tag.text)
    
tags

['change', 'deep-thoughts', 'thinking', 'world']

### Our data:

In [16]:
print('text:', text)
print('author name: ', author_name)
print('author link: ', author_link)
print('tags: ', tags)

text: “The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
author name:  Albert Einstein
author link:  http://quotes.toscrape.com/author/Albert-Einstein
tags:  ['change', 'deep-thoughts', 'thinking', 'world']


# Step 7
> We create a function to make our code reusable.

Now that we know how to collect this data from a quote div, we can compile the code into a [function](https://www.geeksforgeeks.org/functions-in-python/) called `quote_data`. This allows us to grab a quote div, feed it into the function like so...
> `quote_data(quote_div)`

...and receive all of the data from that div.

In [17]:
def quote_data(quote_div):
    
    # Collect the quote
    text = quote_div.find('span', {'class':'text'})
    text = text.text
    
    # Collect the author name
    author = quote_div.find('small', {'class': 'author'})
    author_name = author.text
    
    # Collect author link
    author_link = author.findNextSibling().attrs.get('href') #.get means code continues to run even despite errors
    author_link = url + author_link
    
    # Collect tags
    tag_container = quote_div.find('div', {'class': 'tags'})
    tag_links = tag_container.find_all('a')

    tags = []
    for tag in tag_links:
        tags.append(tag.text)
        
    # Return data as a dictionary
    return {'text': text, 'author': author_name, 'bio': author_link, 'tags': tags}

### Let's test our fuction.

In [19]:
quote_data(quote_divs[4])

{'text': "“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”",
 'author': 'Marilyn Monroe',
 'bio': 'http://quotes.toscrape.com/author/Marilyn-Monroe',
 'tags': ['be-yourself', 'inspirational']}

Now we can collect the data from every quote on the first page with a simple [```for loop```](https://www.w3schools.com/python/python_for_loops.asp)!

In [21]:
page_one_data = []
for div in quote_divs:
    # Apply our function on each quote
    data_from_div = quote_data(div)
    page_one_data.append(data_from_div)
    
print(len(page_one_data), 'quotes scraped!')
page_one_data

10 quotes scraped!


[{'text': '“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”',
  'author': 'Albert Einstein',
  'bio': 'http://quotes.toscrape.com/author/Albert-Einstein',
  'tags': ['change', 'deep-thoughts', 'thinking', 'world']},
 {'text': '“It is our choices, Harry, that show what we truly are, far more than our abilities.”',
  'author': 'J.K. Rowling',
  'bio': 'http://quotes.toscrape.com/author/J-K-Rowling',
  'tags': ['abilities', 'choices']},
 {'text': '“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”',
  'author': 'Albert Einstein',
  'bio': 'http://quotes.toscrape.com/author/Albert-Einstein',
  'tags': ['inspirational', 'life', 'live', 'miracle', 'miracles']},
 {'text': '“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”',
  'author': 'Jane Austen',
  'bio': 'http://quotes.toscrape.com/author

# We just scraped an entire webpage!

![](https://media.giphy.com/media/KiXl0vfc9XIIM/giphy.gif)

### Level Up: What if we wanted to scrape the quotes from *every* page?

**Step 1:** The first thing we do is take the code from above that scraped the data for all of the quotes on page one, and move it into a function called `scrape_page`.

In [22]:
def scrape_page(quote_divs):
    data = []
    for div in quote_divs:
        div_data = quote_data(div)
        data.append(div_data)
        
    return data

**Step 2:** We grab the code we used at the very beginning to collect the html and the list of divs from a web page.

In [23]:
base_url = 'http://quotes.toscrape.com'
response = requests.get(url)
html = response.text
soup = BeautifulSoup(html, 'lxml')
quote_divs = soup.find_all('div', {'class': 'quote'})

**Step 3:** We feed all of the `quote_divs` into our newly made `parse_page` function to grab all of the data from that page.

In [24]:
data = scrape_page(quote_divs)

In [25]:
data[:2]

[{'text': '“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”',
  'author': 'Albert Einstein',
  'bio': 'http://quotes.toscrape.com/author/Albert-Einstein',
  'tags': ['change', 'deep-thoughts', 'thinking', 'world']},
 {'text': '“It is our choices, Harry, that show what we truly are, far more than our abilities.”',
  'author': 'J.K. Rowling',
  'bio': 'http://quotes.toscrape.com/author/J-K-Rowling',
  'tags': ['abilities', 'choices']}]

**Step 4:** We check to see if there is a `Next Page` button at the bottom of the page.

*This requires multiple steps.*

1. We grab the outer container that has a class of `pager`.

In [ ]:
pager = soup.find('ul', {'class': 'pager'})

If there is no pager element on the webpage, pager will be set to `None`.

2. We use an if check to make sure a pager exists:

In [ ]:
if pager:
    next_page = pager.find('li', {'class': 'next'})

3. We then check to see if a `Next Page` button exists on the page. 

    - Every page has a next button except the *last* page which only has a `Previous Page` button. Basically, we're checking to see if the `Next button` exists. It it does, we "click" it.

In [ ]:
if next_page:
    next_page = next_page.findChild('a')\
                         .attrs\
                         .get('href')

With most webscraping tools, "clicking a button" means collecting the link inside the button and making a new request.

If a link is pointing to a page on the same website, the links are usually just the forward slashes that need to be added to the base website url. This is called a `relative` link.

**Step 5:** Collect the relative link that points to the next page, and add it to our base_url

In [ ]:
next_page = url + next_page
next_page

**Step 6:** We repeat the exact same process for this new link!

ie:
1. Make request using a url that points to the next page.
2. Scrape quote divs
3. Collect data from every quote div on that page
4. Find the `Next page` button.
5. Collect the url from the button
6. Repeat

So how do we do this over and over again without repeating ourselves?

The first step is compile all of these steps into a new function called `scrape_quotes`.

The second step is, something called `recursion`. 

<center><h1><u>Recursion</u></h1></center>

![](https://media.giphy.com/media/l1J9R1Q7LJGSZOxFe/giphy.gif)

> **Recursion** is a bit of a mind bend, so don't feel bad if it is hard to wrap your brain around. It took me a while to be able to understand recursive functions!

Essentially, recursion is where we use a function inside of itself.

**In this instance,** our code will be telling the computer, if there is a `Next page` button, rerun all of the code again on the page the next button points us to and check to see if there is a `Next page` button on *that* page. If there is, keep repeating the process until a `Next page` button is not found.

In [ ]:
def scrape_quotes(url):
    base_url = 'http://quotes.toscrape.com'
    response = requests.get(url)
    html = response.text
    soup = BeautifulSoup(html, 'lxml')
    quote_divs = soup.find_all('div', {'class': 'quote'})
    data = scrape_page(quote_divs)
    
    pager = soup.find('ul', {'class': 'pager'})
    if pager:
        next_page = pager.find('li', {'class': 'next'})
        
        if next_page:
            next_page =  next_page.findChild('a')\
                                  .attrs\
                                  .get('href')
            
            next_page = base_url + next_page
            print('Scraping', next_page)
            ## This is where the recursion happens
            data += scrape_quotes(next_page)
   
    return data

Now we can set a variable called `data` that is the output of our recursive function!

> A print statement has been added to output what page is being scraped

In [ ]:
data = scrape_quotes(url)
print(len(data), 'Quotes scraped!')

Now we can visualize and explore our data!

In [ ]:
def count_tags(quote):
    return len(quote['tags'])

def tag_lengths(data):
    lengths = []
    for quote in data:
        lengths.append(count_tags(quote))
        
    return lengths
        
lengths = tag_lengths(data)
plt.figure(figsize=(10,6))   
plt.hist(lengths, bins=9)
plt.title('Number of Tags');

# Saving our data

There are multiple ways to save data to a file. Pandas, `The Excel of Python` allows us to do this easily.

In [ ]:
df = pd.DataFrame(data)
df.head()

In [ ]:
df.to_csv('quotes_data.csv', index=False)

## Scraping Tables with Pandas

In [ ]:
url = 'https://en.wikipedia.org/wiki/List_of_current_United_States_senators'
pd.read_html(url, attrs={'id': 'senators'})[0].head()

Pandas `read_html` is a really handy tool, but it isn't perfect. You'll notice that the resulting table above is quite messy. The urls for the portrait column are replaced 'NaN' values, the Born column is returning the age of the senator instead of their birth date, the occupation and previous office columns are lists inside strings with the spacing between letters removed. For this table, if we wanted to take full advantage of the available data, we would need to scrape this table by hand.  

To do that, we can begin by isolating the table, and then isolating the data rows in the table.

Below is a function that returns a cleaned version of the senators table.

In [ ]:
def scrape_senate_table():
    # Connect to the wikipedia page
    response = requests.get('https://en.wikipedia.org/wiki/List_of_current_United_States_senators')
    # Collect the html from the page
    html = response.text
    # Parse the html with BeautifulSoup
    soup = BeautifulSoup(html)
    # Find the senators table
    table = soup.find('table', {'id': 'senators'})
    
    # The first row in the table contains the column names.
    # Isolate the first row
    rows = table.find_all('tr')[1:]
    # Create an empty list to append the row data to
    senate_data = []
    # Create some cleaning functions for text data
    remove_new_line = lambda x: x.replace('\n', '')
    split_new_line = lambda x: x.split('\n')
    
    # Loop over each row
    for row in rows:
        # find all td tags from the row
        td_tags = row.find_all('td')
        # The senators table merges the cell for `state name` so it spans both 
        # senators from a state. 
        # When parsing the html, the state name only appears for the senator
        # that appears first in the table, and the second appearing senator has 
        # one less td tag.
        # Because of this we need to check the length of the td tags
        # If there is one less tag, we use the state name from the previous iteration
        if len(td_tags) == 10:
            previous_element = [td_tags[0]]
        else:
            previous_element += td_tags
            td_tags = previous_element
        
        # Collect the state name
        state = remove_new_line(td_tags[0].text)
        # Collect the image url
        image = td_tags[1].find('img').attrs['src']
        # Collect the name of the senator
        name = remove_new_line(row.find('th').text)
        # Collect the css color string for the political party
        party_color = td_tags[2].attrs['style'].split(':')[-1]
        # Collect the party name
        party_name = remove_new_line(td_tags[3].text)
        # Collect the date of birth
        dob = ' '.join(remove_new_line(td_tags[4]\
                                       .text)\
                                       .strip()\
                                       .split(' ')[1:4])
        # Collect the occupation
        occupation = split_new_line(BeautifulSoup(str(td_tags[5])\
                                                  .replace('<br/>', '\n'))\
                                                  .td\
                                                  .text)
        # If only one occupation is present
        # Pull that occupation out of the list
        # And return a single string
        if occupation[1] == '':
            occupation = occupation[0]
        # Collect the previous office
        previous_office = split_new_line(BeautifulSoup(str(td_tags[6])\
                                                       .replace('<br/>', '\n'))\
                                                       .td\
                                                      .text)
        # If only one previous office is present
        # Pull that value out of the list
        # And return a single string
        if previous_office[1] == '':
            previous_office = previous_office[0]
        # Collect assumed office
        assumed_office = remove_new_line(td_tags[7].text)
        # Collect the end of their term
        term_up = remove_new_line(td_tags[8].text)
        # Collect their residence
        residence = split_new_line(td_tags[9].text)
        
        # If only one residence is present
        # Pull that value out of the list
        # And return a single string
        if residence[1] == '':
            residence = residence[0]
            # Many of the residences have a 
            # link pointing to additional information
            # We do not need this
            if '[' in residence:
                residence = residence[:-3]
        # Create data dictionary
        collected = {'state': state, 'name': name,'dob': dob, 'party': party_name,
                    'party_color': party_color,
                    'occupation': occupation,
                    'previous_office': previous_office,
                    'assumed_office': assumed_office,
                    'term_up': term_up, 'residence': residence,
                    'portrait': image}
        # Append dictionary to the senate_data list
        senate_data.append(collected)
        
    # Once data from all rows has been collected
    # return the data as a pandas dataframe
    return pd.DataFrame(senate_data)

In [ ]:
scrape_senate_table().head(3)

# Web scraping is a powerful tool. 

It can be used:
- To discover the most in demand skills for thousands of online job postings.
- To learn the average price of a product from thousands of online sale.
- To research social media networks.

**And so much more!** As our world continues to develop online markets and communities, the uses for webscraping continue to grow. In the end, the power of web scraping comes from the ability to create datasets that otherwise do not exist, or at the very least, are not readily available to the public.
